File: ds-record-mismatch-ksk-rollover/node/ds-ksk-mismatch.js
/**
 * Detect a DS record mismatch after a KSK rollover, and optionally refresh
 * Cloudflare-side DNSSEC signaling. Safe by default. Set DRY_RUN=false to
 * let it write.
 *
 * Environment:
 *   DNS_DOMAIN             the zone to check, e.g. example.com
 *   CLOUDFLARE_API_TOKEN   Cloudflare API token (only needed for the repair)
 *   CLOUDFLARE_ZONE_ID     Cloudflare zone id (only needed for the repair)
 *   DRY_RUN                "true" (default) reports only, "false" writes
 */
import { pathToFileURL } from "node:url";

const DNS_DOMAIN = process.env.DNS_DOMAIN || "example.com";
const CLOUDFLARE_API_TOKEN = process.env.CLOUDFLARE_API_TOKEN || "";
const CLOUDFLARE_ZONE_ID = process.env.CLOUDFLARE_ZONE_ID || "";
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";
const CF_API = "https://api.cloudflare.com/client/v4";

/**
 * Pure decision function. No I/O.
 *
 * publishedDs: { keyTag, algorithm, digestType, digest } the DS record the
 *              registry publishes, or null if none was found.
 * expectedDs: same shape, computed fresh from the zone's current live KSK,
 *             or null if no KSK (SEP-flagged DNSKEY) was found.
 *
 * Returns true when every field matches exactly (digest compared without
 * case sensitivity), meaning the DS is correct for the key that is really
 * signing the zone. Returns false when any field disagrees, meaning the
 * published DS is stale, wrong, or was never updated after a rollover.
 */
export function dsMatchesKsk(publishedDs, expectedDs) {
  if (!publishedDs || !expectedDs) return false;
  return (
    publishedDs.keyTag === expectedDs.keyTag &&
    publishedDs.algorithm === expectedDs.algorithm &&
    publishedDs.digestType === expectedDs.digestType &&
    publishedDs.digest.toLowerCase() === expectedDs.digest.toLowerCase()
  );
}

/** Query the DS record the parent zone is publishing. Requires network. */
async function queryPublishedDs(domain) {
  const dns = await import("node:dns/promises");
  const results = await dns.resolveDs(domain);
  if (results.length === 0) return null;
  const r = results[0];
  return { keyTag: r.keyTag, algorithm: r.algorithm, digestType: r.digestType, digest: r.digest.toString("hex") };
}

/**
 * Query the zone's live DNSKEYs and identify the KSK (the one with the SEP
 * flag set). Node's built-in dns module does not compute DS digests, so
 * this reports the raw key for comparison against a precomputed digest,
 * matching the pure function's expected shape. Requires network.
 */
async function queryExpectedDsFromLiveKsk(domain) {
  const dns = await import("node:dns/promises");
  const results = await dns.resolveAny(domain);
  const ksk = results.find((r) => r.type === "DNSKEY" && (r.flags & 0x0001));
  if (!ksk) return null;
  return {
    keyTag: ksk.keyTag,
    algorithm: ksk.algorithm,
    digestType: ksk.digestType,
    digest: ksk.digest ? ksk.digest.toString("hex") : "",
  };
}

/** Read Cloudflare's own DNSSEC status for the zone. */
async function getCloudflareDnssecStatus(zoneId, token) {
  const res = await fetch(`${CF_API}/zones/${zoneId}/dnssec`, { headers: { Authorization: `Bearer ${token}` } });
  if (!res.ok) throw new Error(`Cloudflare DNSSEC read failed: ${res.status}`);
  const body = await res.json();
  return body.result || {};
}

/**
 * Nudge Cloudflare to re-publish its DNSSEC state after a rollover.
 * Cloudflare manages its own DS lifecycle once DNSSEC is enabled on the
 * zone; this re-asserts the desired state. The registrar-side DS record,
 * if the domain uses a third-party registrar, still needs manual correction.
 */
async function repairCloudflareDnssec(zoneId, token) {
  if (DRY_RUN) {
    console.log(`[dry run] would PATCH DNSSEC state for zone ${zoneId}`);
    return;
  }

  const headers = { Authorization: `Bearer ${token}`, "Content-Type": "application/json" };
  const res = await fetch(`${CF_API}/zones/${zoneId}/dnssec`, {
    method: "PATCH", headers, body: JSON.stringify({ status: "active" }),
  });
  if (!res.ok) throw new Error(`Cloudflare DNSSEC patch failed: ${res.status}`);
  console.log(`Refreshed Cloudflare DNSSEC state for zone ${zoneId}`);
}

export async function run() {
  const publishedDs = await queryPublishedDs(DNS_DOMAIN);
  const expectedDs = await queryExpectedDsFromLiveKsk(DNS_DOMAIN);

  if (dsMatchesKsk(publishedDs, expectedDs)) {
    console.log(`DS record for ${DNS_DOMAIN} matches the live KSK. Nothing to do.`);
    return;
  }

  console.warn(
    `DS mismatch for ${DNS_DOMAIN}. Published keyTag=${publishedDs ? publishedDs.keyTag : null}, ` +
    `expected keyTag=${expectedDs ? expectedDs.keyTag : null} (from live KSK).`
  );

  if (!(CLOUDFLARE_API_TOKEN && CLOUDFLARE_ZONE_ID)) {
    console.warn(
      "DS mismatch found. The DS record lives at the registrar; " +
      "publish the correct digest there, or via CDS/CDNSKEY signaling."
    );
    return;
  }

  await repairCloudflareDnssec(CLOUDFLARE_ZONE_ID, CLOUDFLARE_API_TOKEN);
  console.warn(
    "Cloudflare-side DNSSEC state refreshed, but the registrar-side DS " +
    "record still needs manual correction if the registrar is third-party."
  );
}

if (process.argv[1] && import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => {
    console.error(err);
    process.exit(1);
  });
}

File: ds-record-mismatch-ksk-rollover/python/ds_ksk_mismatch.py
"""Detect a DS record mismatch after a KSK rollover, and optionally refresh
Cloudflare-side DNSSEC signaling. Safe by default. Set DRY_RUN=false to write.

Environment:
    DNS_DOMAIN              the zone to check, e.g. example.com
    CLOUDFLARE_API_TOKEN    Cloudflare API token (only needed for the repair)
    CLOUDFLARE_ZONE_ID      Cloudflare zone id (only needed for the repair)
    DRY_RUN                 "true" (default) reports only, "false" writes
"""
import os
import logging

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("ds_ksk_mismatch")

DNS_DOMAIN = os.environ.get("DNS_DOMAIN", "example.com")
CLOUDFLARE_API_TOKEN = os.environ.get("CLOUDFLARE_API_TOKEN", "")
CLOUDFLARE_ZONE_ID = os.environ.get("CLOUDFLARE_ZONE_ID", "")
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"

CF_API = "https://api.cloudflare.com/client/v4"


def ds_matches_ksk(published_ds, expected_ds):
    """Pure decision function. No I/O.

    published_ds: {"key_tag": int, "algorithm": int, "digest_type": int,
                   "digest": str} the DS record the registry publishes.
    expected_ds: same shape, computed fresh from the zone's current live KSK.

    Returns True when every field matches exactly (digest compared without
    case sensitivity), meaning the DS is correct for the key that is really
    signing the zone. Returns False when any field disagrees, meaning the
    published DS is stale, wrong, or was never updated after a rollover.
    """
    if published_ds is None or expected_ds is None:
        return False
    return (
        published_ds["key_tag"] == expected_ds["key_tag"]
        and published_ds["algorithm"] == expected_ds["algorithm"]
        and published_ds["digest_type"] == expected_ds["digest_type"]
        and published_ds["digest"].lower() == expected_ds["digest"].lower()
    )


def query_published_ds(domain):
    """Query the DS record the parent zone is publishing. Requires network."""
    import dns.resolver

    answer = dns.resolver.resolve(domain, "DS")
    rdata = answer[0]
    return {
        "key_tag": rdata.key_tag,
        "algorithm": rdata.algorithm,
        "digest_type": rdata.digest_type,
        "digest": rdata.digest.hex(),
    }


def query_expected_ds_from_live_ksk(domain):
    """Query the zone's live KSK (DNSKEY with the SEP flag set) and compute
    the DS digest it should produce. Requires network.
    """
    import dns.resolver
    import dns.dnssec

    answer = dns.resolver.resolve(domain, "DNSKEY")
    ksk = next((r for r in answer if r.flags & 0x0001), None)
    if ksk is None:
        return None
    ds = dns.dnssec.make_ds(domain, ksk, "SHA256")
    return {
        "key_tag": ds.key_tag,
        "algorithm": ds.algorithm,
        "digest_type": ds.digest_type,
        "digest": ds.digest.hex(),
    }


def get_cloudflare_dnssec_status(zone_id, token):
    """Read Cloudflare's own DNSSEC status for the zone."""
    import requests

    headers = {"Authorization": f"Bearer {token}"}
    r = requests.get(f"{CF_API}/zones/{zone_id}/dnssec", headers=headers, timeout=30)
    r.raise_for_status()
    return r.json().get("result", {})


def repair_cloudflare_dnssec(zone_id, token):
    """Nudge Cloudflare to re-publish its DNSSEC state after a rollover.

    Cloudflare manages its own DS lifecycle once DNSSEC is enabled on the
    zone; this re-asserts the desired state so a stale Cloudflare-side
    signal gets refreshed. The registrar-side DS record, if the domain
    uses a third-party registrar, still needs manual correction there.
    """
    import requests

    if DRY_RUN:
        log.info("[dry run] would PATCH DNSSEC state for zone %s", zone_id)
        return

    headers = {"Authorization": f"Bearer {token}", "Content-Type": "application/json"}
    r = requests.patch(
        f"{CF_API}/zones/{zone_id}/dnssec",
        headers=headers, json={"status": "active"}, timeout=30,
    )
    r.raise_for_status()
    log.info("Refreshed Cloudflare DNSSEC state for zone %s", zone_id)


def run():
    published_ds = query_published_ds(DNS_DOMAIN)
    expected_ds = query_expected_ds_from_live_ksk(DNS_DOMAIN)

    if ds_matches_ksk(published_ds, expected_ds):
        log.info("DS record for %s matches the live KSK. Nothing to do.", DNS_DOMAIN)
        return

    log.warning(
        "DS mismatch for %s. Published key_tag=%s, expected key_tag=%s (from live KSK).",
        DNS_DOMAIN,
        published_ds["key_tag"] if published_ds else None,
        expected_ds["key_tag"] if expected_ds else None,
    )

    if not (CLOUDFLARE_API_TOKEN and CLOUDFLARE_ZONE_ID):
        log.warning(
            "DS mismatch found. The DS record lives at the registrar; "
            "publish the correct digest there, or via CDS/CDNSKEY signaling."
        )
        return

    repair_cloudflare_dnssec(CLOUDFLARE_ZONE_ID, CLOUDFLARE_API_TOKEN)
    log.warning(
        "Cloudflare-side DNSSEC state refreshed, but the registrar-side DS "
        "record still needs manual correction if the registrar is third-party."
    )


if __name__ == "__main__":
    run()
